charon/src/transform/ullbc_to_llbc.rs

```diff
@@ -856,117 +872,86 @@ fn compute_switch_exits_explore(
-    // 2. in this intersection, we take the first block id (remember we use
-    //   topological sort), which will be our exit node.
-    //
-    // The reason behind 1 is that some branches of a match can join themselves,
-    // before joining other branches. For example:
-    // ```
-    // let y = match x {
-    //   | E1 | E2 => 0, // Those 2 branches lead to the same node
-    //   | E3 => 1,
-    // };
-    // // But the 3 branches join this point: this is the proper exit
-    // return y;
-    // ```
-    //
-    // Note that we're definitely not looking for performance here (and that
-    // there shouldn't be too many blocks in a function body), but rather
-    // quality of the generated code. If the following works well but proves
-    // to be too slow, we'll think of a way of making it faster.
-    // Note: actually, we could look only for *any* two pair of children
-    // whose successors intersection is non empty: I think it works in the
-    // general case.
-    else {
-        let mut max_number_inter: u32 = 0;
-        let mut max_inter_succs: BTreeSet<OrdBlockId> = BTreeSet::new();
-
-        // For every child
-        for (i, mut i_succs) in children_succs.iter().cloned().enumerate() {
-            let mut current_number_inter = 1;
-            // Note that we need to add the children themselves to the
-            // sets of successors
-            i_succs.insert(make_ord_block_id(children[i], tsort_map));
-            let mut current_inter_succs: BTreeSet<OrdBlockId> = i_succs;
-
-            // Compute the "best" intersection with all the other children
-            for (j, mut j_succs) in children_succs.iter().cloned().enumerate() {
-                j_succs.insert(make_ord_block_id(children[j], tsort_map));
-
-                // Annoying that we have to clone the current intersection set...
-                let inter: BTreeSet<OrdBlockId> = current_inter_succs
-                    .intersection(&j_succs)
-                    .copied()
-                    .collect();
-
-                if !inter.is_empty() {
-                    current_number_inter += 1;
-                    current_inter_succs = inter;
-                }
-            }
+    // Compute the successors
+    // Add the children themselves in their sets of successors
+    let all_succs: BTreeSet<OrdBlockId> = children.iter().fold(BTreeSet::new(), |acc, s| {
+        // Add the successors to the accumulator
+        let mut acc: BTreeSet<_> = acc.union(&s.succs).copied().collect();
+        // Add the child itself
+        acc.insert(make_ord_block_id(s.id, tsort_map));
+        acc
+    });
 
-            // Update the best intersection, if necessary
-            if current_number_inter > max_number_inter {
-                max_number_inter = current_number_inter;
-                max_inter_succs = current_inter_succs;
+    // Compute the "flows" (if there are children)
+    // TODO: this is computationally expensive...
```

Comment:
That should be linear in the size of the graph, no?

Reply:
This is linear *for each node*.
And the size of the fraction is also linear in the total size of the graph...

Reply:
Can you not traverse the DAG in topological order to compute the flow in linear time in the size of the graph?

Reply:
The flow at a given node unfortunately depends on the node on which you start:
```
A
|  \
B C
|   |
|  D
|  /
E
```
The flow at `D` is 0.5 if you start from `A` and `1` if you start from `D`. So in the end we need to remember `V^2` values though there is probably a smart way of not remembering all values).

Reply:
For instance, I guess what's really important is the flow at the join nodes (i.e., starting from `C` we don't need to remember the flow at `D`, what really matters is the flow at `E` as there is more than one incoming edge in `E`).

Reply:
Ah I see, we need to compute it once for each entry node, right? I understand the comment now.

Reply:
Yes, exactly